File: frontend/src/api/credit.ts
import { API_ENDPOINTS } from '../config/endpoints'
import type { CustomerStatus } from '../types/customer'
import { apiGet, apiPost } from './client'

function isRecord(x: unknown): x is Record<string, unknown> {
  return typeof x === 'object' && x !== null
}

/** Tolerate camelCase or missing fields from proxies / older servers. */
export function normalizeCreditSummary(raw: unknown): CreditSummaryResponse {
  if (!isRecord(raw)) return { total_outstanding: 0, debtor_count: 0 }
  const to = raw.total_outstanding ?? raw.totalOutstanding ?? 0
  const dc = Number(raw.debtor_count ?? raw.debtorCount ?? 0)
  return {
    total_outstanding: typeof to === 'number' || typeof to === 'string' ? to : 0,
    debtor_count: Number.isFinite(dc) ? dc : 0,
  }
}

export function normalizeDebtorRows(raw: unknown): CreditDebtorRow[] {
  if (!Array.isArray(raw)) return []
  const out: CreditDebtorRow[] = []
  for (const item of raw) {
    if (!isRecord(item)) continue
    const id = Number(item.id)
    if (!Number.isFinite(id) || id < 1) continue
    const name = String(item.name ?? '')
    const phone = String(item.phone ?? '')
    const debt = item.debt_balance ?? item.debtBalance ?? 0
    const cl = item.credit_limit ?? item.creditLimit
    const status = (item.status ?? 'active') as CustomerStatus
    out.push({
      id,
      name,
      phone,
      debt_balance: typeof debt === 'number' || typeof debt === 'string' ? debt : 0,
      credit_limit:
        cl === null || cl === undefined || cl === '' ? null : (cl as string | number),
      status,
    })
  }
  return out
}

export function normalizeCreditPaymentRows(raw: unknown): CreditPaymentListItem[] {
  if (!Array.isArray(raw)) return []
  const out: CreditPaymentListItem[] = []
  for (const item of raw) {
    if (!isRecord(item)) continue
    const id = Number(item.id)
    const customer_id = Number(item.customer_id ?? item.customerId)
    if (!Number.isFinite(id) || !Number.isFinite(customer_id)) continue
    out.push({
      id,
      customer_id,
      customer_name: String(item.customer_name ?? item.customerName ?? ''),
      amount: (item.amount ?? 0) as string | number,
      payment_method: String(item.payment_method ?? item.paymentMethod ?? ''),
      note: String(item.note ?? ''),
      balance_after: (item.balance_after ?? item.balanceAfter ?? 0) as string | number,
      cash_register_session_id: (() => {
        const raw = item.cash_register_session_id ?? item.cashRegisterSessionId
        if (raw == null) return null
        const n = Number(raw)
        return Number.isFinite(n) && n >= 1 ? n : null
      })(),
      created_at: String(item.created_at ?? item.createdAt ?? ''),
    })
  }
  return out
}

export type CreditSummaryResponse = {
  total_outstanding: string | number
  debtor_count: number
}

export type CreditDebtorRow = {
  id: number
  name: string
  phone: string
  debt_balance: string | number
  credit_limit: string | number | null
  status: CustomerStatus
}

export type CreditPaymentListItem = {
  id: number
  customer_id: number
  customer_name: string
  amount: string | number
  payment_method: string
  note: string
  balance_after: string | number
  cash_register_session_id: number | null
  created_at: string
}

export type CreditLedgerEntryApi = {
  kind: 'charge' | 'payment'
  record_id: number
  created_at: string
  amount: string | number
  balance_after: string | number
  sale_id: number | null
  payment_method: string | null
  note: string | null
}

export type CustomerCreditLedgerResponse = {
  customer_id: number
  entries: CreditLedgerEntryApi[]
}

export type CreditPaymentCreatePayload = {
  customer_id: number
  amount: string
  payment_method: 'cash' | 'moncash' | 'transfer'
  note: string
  cash_register_session_id: number | null
}

export type CreditPaymentResponse = Omit<CreditPaymentListItem, 'customer_name'>

export function buildCreditPaymentsListUrl(params: { skip?: number; limit?: number } = {}): string {
  const sp = new URLSearchParams()
  if (params.skip != null) sp.set('skip', String(params.skip))
  if (params.limit != null) sp.set('limit', String(params.limit))
  const qs = sp.toString()
  return qs ? `${API_ENDPOINTS.credit.payments}?${qs}` : API_ENDPOINTS.credit.payments
}

export async function getCreditSummaryApi(): Promise<CreditSummaryResponse> {
  const raw = await apiGet<unknown>(API_ENDPOINTS.credit.summary)
  return normalizeCreditSummary(raw)
}

export async function listCreditDebtorsApi(): Promise<CreditDebtorRow[]> {
  const raw = await apiGet<unknown>(API_ENDPOINTS.credit.debtors)
  return normalizeDebtorRows(raw)
}

export async function listCreditPaymentsApi(params: {
  skip?: number
  limit?: number
} = {}): Promise<CreditPaymentListItem[]> {
  const raw = await apiGet<unknown>(buildCreditPaymentsListUrl(params))
  return normalizeCreditPaymentRows(raw)
}

export async function getCustomerCreditLedgerApi(customerId: number): Promise<CustomerCreditLedgerResponse> {
  return apiGet<CustomerCreditLedgerResponse>(API_ENDPOINTS.credit.ledger(customerId))
}

export async function createCreditPaymentApi(body: CreditPaymentCreatePayload): Promise<CreditPaymentResponse> {
  return apiPost<CreditPaymentResponse>(API_ENDPOINTS.credit.payments, body)
}

File: backend/src/routers/credit.py
from __future__ import annotations

from decimal import Decimal

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import and_, func, or_, select
from sqlalchemy.orm import Session

from src.auth.deps import get_current_user
from src.db.session import get_db
from src.models.customer import Customer
from src.models.customer_credit_payment import CustomerCreditPayment
from src.models.user import User
from src.schemas.credit import (
    CreditLedgerEntry,
    CreditPaymentCreate,
    CreditPaymentListItem,
    CreditPaymentResponse,
    CreditSummaryResponse,
    CustomerCreditLedgerResponse,
    DebtorRow,
)
from src.services.credit_service import (
    CreditWorkflowError,
    CustomerNotFoundForCreditError,
    build_customer_ledger,
    record_credit_payment,
)

router = APIRouter(prefix="/credit", tags=["credit"])


def _credit_http_error(exc: CreditWorkflowError) -> HTTPException:
    if isinstance(exc, CustomerNotFoundForCreditError):
        return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    return HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc))


@router.get("/summary", response_model=CreditSummaryResponse)
def credit_summary(
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
) -> CreditSummaryResponse:
    debt_positive = func.coalesce(Customer.debt_balance, 0) > 0
    total = db.scalar(
        select(func.coalesce(func.sum(Customer.debt_balance), 0)).where(debt_positive)
    )
    count = db.scalar(select(func.count()).select_from(Customer).where(debt_positive))
    return CreditSummaryResponse(
        total_outstanding=Decimal(total or 0),
        debtor_count=int(count or 0),
    )


@router.get("/debtors", response_model=list[DebtorRow])
def list_debtors(
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
    skip: int = Query(default=0, ge=0),
    limit: int = Query(default=200, ge=1, le=500),
) -> list[DebtorRow]:
    debt_col = func.coalesce(Customer.debt_balance, 0)
    stmt = (
        select(Customer)
        .where(
                debt_col > 0,
        )
        .order_by(debt_col.desc(), Customer.name.asc())
        .offset(skip)
        .limit(limit)
    )
    rows = db.scalars(stmt).all()
    return [
        DebtorRow(
            id=c.id,
            name=c.name,
            phone=c.phone,
            debt_balance=c.debt_balance,
            credit_limit=c.credit_limit,
            status=c.status,
        )
        for c in rows
    ]


@router.get("/payments", response_model=list[CreditPaymentListItem])
def list_credit_payments(
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
    skip: int = Query(default=0, ge=0),
    limit: int = Query(default=50, ge=1, le=300),
) -> list[CreditPaymentListItem]:
    stmt = (
        select(CustomerCreditPayment, Customer.name)
        .join(Customer, Customer.id == CustomerCreditPayment.customer_id)
        .order_by(CustomerCreditPayment.created_at.desc(), CustomerCreditPayment.id.desc())
        .offset(skip)
        .limit(limit)
    )
    rows = db.execute(stmt).all()
    out: list[CreditPaymentListItem] = []
    for p, customer_name in rows:
        out.append(
            CreditPaymentListItem(
                id=p.id,
                customer_id=p.customer_id,
                amount=p.amount,
                payment_method=p.payment_method,
                note=p.note,
                balance_after=p.balance_after,
                cash_register_session_id=p.cash_register_session_id,
                created_at=p.created_at,
                customer_name=customer_name,
            )
        )
    return out


@router.get("/customers/{customer_id}/ledger", response_model=CustomerCreditLedgerResponse)
def get_customer_credit_ledger(
    customer_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
) -> CustomerCreditLedgerResponse:
    if db.get(Customer, customer_id) is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Customer not found")
    raw = build_customer_ledger(db, customer_id)
    entries = [
        CreditLedgerEntry(
            kind=e["kind"],  # type: ignore[arg-type]
            record_id=e["record_id"],
            created_at=e["created_at"],
            amount=e["amount"],
            balance_after=e["balance_after"],
            sale_id=e["sale_id"],
            payment_method=e["payment_method"],
            note=e["note"],
        )
        for e in raw
    ]
    return CustomerCreditLedgerResponse(customer_id=customer_id, entries=entries)


@router.post("/payments", response_model=CreditPaymentResponse, status_code=status.HTTP_201_CREATED)
def create_credit_payment(
    payload: CreditPaymentCreate,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
) -> CreditPaymentResponse:
    try:
        result = record_credit_payment(
            db,
            customer_id=payload.customer_id,
            amount=payload.amount,
            payment_method=payload.payment_method,
            note=payload.note,
            cash_register_session_id=payload.cash_register_session_id,
        )
    except CreditWorkflowError as e:
        raise _credit_http_error(e) from e

    p = result.payment
    return CreditPaymentResponse(
        id=p.id,
        customer_id=p.customer_id,
        amount=p.amount,
        payment_method=p.payment_method,
        note=p.note,
        balance_after=p.balance_after,
        cash_register_session_id=p.cash_register_session_id,
        created_at=p.created_at,
    )
